File: day1/main.py
from collections import defaultdict


def main():
    part_one()
    part_two()


def part_one():
    current_frequency = 0

    input_file = open('input.txt', 'r')

    for current_line in input_file:
        frequency = int(current_line)
        current_frequency += frequency

    input_file.close()

    print 'Resulting Frequency is ' + str(current_frequency)


def part_two():
    current_frequency = 0

    past_frequencies_dic = defaultdict(bool)

    past_frequencies_dic[0] = True

    with open('input.txt', 'r') as input_file:
        lines = [int(line) for line in input_file]

    frequency_twice_not_found = True

    while frequency_twice_not_found:
        for current_line in lines:
            frequency = int(current_line)
            current_frequency += frequency
            if past_frequencies_dic[current_frequency]:
                print str(current_frequency) + ' appeared twice'
                frequency_twice_not_found = False
                break
            else:
                past_frequencies_dic[current_frequency] = True


if __name__ == '__main__':
    main()

File: day1/Main.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Scanner;

public class Main {
    public static void main(String args[]) {
        partOne();
        partTwo();
    }

    private static void partOne() {
        int currentFrequency = 0;

        File inputFile = new File("input.txt");
        try {
            Scanner scanner = new Scanner(inputFile);

            while (scanner.hasNext()) {
                int frequency = scanner.nextInt();
                currentFrequency += frequency;
            }

            scanner.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }

        System.out.println("Resulting Frequency is " + currentFrequency);
    }

    private static void partTwo() {
        int currentFrequency = 0;

        HashMap<Integer, Boolean> pastFrequences = new HashMap<>();

        pastFrequences.put(0, true);

        List<Integer> frequencies = new ArrayList<>();

        File inputFile = new File("input.txt");
        try {
            Scanner scanner = new Scanner(inputFile);

            while (scanner.hasNext()) {
                int frequency = scanner.nextInt();
                frequencies.add(frequency);
            }
            scanner.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }

        boolean twiceFrequencyNotFound = true;

        while (twiceFrequencyNotFound) {
            for (int frequency : frequencies) {
                currentFrequency += frequency;
                boolean isPastFrequency = pastFrequences.containsKey(currentFrequency);
                if (isPastFrequency) {
                    System.out.println(currentFrequency + " appeared twice");
                    twiceFrequencyNotFound = false;
                    break;
                } else {
                    pastFrequences.put(currentFrequency, true);
                }
            }
        }


    }
}
